File: python/src/frtb/market.py
"""Market data containers: zero curves, equity quotes, FX spots.

All bump operations return *new* objects (the engine treats markets as
immutable snapshots) so bump-and-revalue sensitivities cannot leak state.
"""
from __future__ import annotations

import csv
import math
from dataclasses import dataclass, replace
from pathlib import Path
from typing import Dict, List, Sequence


@dataclass(frozen=True)
class Curve:
    """Continuously-compounded zero curve with linear interpolation in tenor.

    Rates are interpolated linearly between nodes and extrapolated flat
    beyond the first/last node.  Discount factor: DF(t) = exp(-z(t) * t).
    """

    tenors: tuple
    rates: tuple

    def __post_init__(self) -> None:
        if len(self.tenors) == 0 or len(self.tenors) != len(self.rates):
            raise ValueError("Curve: tenors and rates must be non-empty and equal length")
        for i in range(1, len(self.tenors)):
            if self.tenors[i] <= self.tenors[i - 1]:
                raise ValueError("Curve: tenors must be strictly increasing")
        for t, r in zip(self.tenors, self.rates):
            if not (math.isfinite(t) and math.isfinite(r)) or t <= 0.0:
                raise ValueError("Curve: tenors must be positive finite, rates finite")

    def rate(self, t: float) -> float:
        """Interpolated zero rate at time t (flat extrapolation)."""
        if not math.isfinite(t) or t < 0.0:
            raise ValueError(f"Curve.rate: invalid time {t}")
        ts, rs = self.tenors, self.rates
        if t <= ts[0]:
            return rs[0]
        if t >= ts[-1]:
            return rs[-1]
        for i in range(1, len(ts)):
            if t <= ts[i]:
                w = (t - ts[i - 1]) / (ts[i] - ts[i - 1])
                return rs[i - 1] * (1.0 - w) + rs[i] * w
        return rs[-1]  # pragma: no cover - unreachable

    def df(self, t: float) -> float:
        """Discount factor exp(-z(t)*t); DF(0) = 1."""
        if t == 0.0:
            return 1.0
        return math.exp(-self.rate(t) * t)

    def bumped_node(self, tenor: float, size: float) -> "Curve":
        """Return a curve with the zero rate at one node shifted by `size` (absolute)."""
        if tenor not in self.tenors:
            raise ValueError(f"Curve.bumped_node: tenor {tenor} is not a curve node")
        rates = tuple(r + size if tt == tenor else r for tt, r in zip(self.tenors, self.rates))
        return Curve(self.tenors, rates)

    def bumped_parallel(self, size: float) -> "Curve":
        """Return a curve with every node shifted by `size` (absolute)."""
        return Curve(self.tenors, tuple(r + size for r in self.rates))


@dataclass(frozen=True)
class EquityQuote:
    """Equity market data: spot, flat lognormal vol, dividend yield, SBM bucket."""

    spot: float
    vol: float
    div_yield: float
    bucket: str

    def __post_init__(self) -> None:
        if not math.isfinite(self.spot) or self.spot <= 0.0:
            raise ValueError(f"EquityQuote: spot must be positive finite, got {self.spot}")
        if not math.isfinite(self.vol) or self.vol < 0.0:
            raise ValueError(f"EquityQuote: vol must be >= 0, got {self.vol}")
        if not math.isfinite(self.div_yield):
            raise ValueError("EquityQuote: div_yield must be finite")


@dataclass(frozen=True)
class Market:
    """Immutable market snapshot: curves per currency, equities per name, FX spots per pair."""

    curves: Dict[str, Curve]
    equities: Dict[str, EquityQuote]
    fx: Dict[str, float]

    def curve(self, ccy: str) -> Curve:
        if ccy not in self.curves:
            raise ValueError(f"Market: no curve for currency '{ccy}'")
        return self.curves[ccy]

    def equity(self, name: str) -> EquityQuote:
        if name not in self.equities:
            raise ValueError(f"Market: no equity quote for '{name}'")
        return self.equities[name]

    def fx_spot(self, pair: str) -> float:
        if pair not in self.fx:
            raise ValueError(f"Market: no FX spot for pair '{pair}'")
        return self.fx[pair]

    # --- bump helpers (all return new Market objects) -------------------
    def with_curve(self, ccy: str, curve: Curve) -> "Market":
        curves = dict(self.curves)
        curves[ccy] = curve
        return Market(curves, self.equities, self.fx)

    def bump_curve_node(self, ccy: str, tenor: float, size: float) -> "Market":
        return self.with_curve(ccy, self.curve(ccy).bumped_node(tenor, size))

    def bump_curve_parallel(self, ccy: str, size: float) -> "Market":
        return self.with_curve(ccy, self.curve(ccy).bumped_parallel(size))

    def bump_equity_spot(self, name: str, rel: float) -> "Market":
        """Relative spot bump: S -> S * (1 + rel)."""
        q = self.equity(name)
        eqs = dict(self.equities)
        eqs[name] = replace(q, spot=q.spot * (1.0 + rel))
        return Market(self.curves, eqs, self.fx)

    def bump_equity_vol(self, name: str, size: float) -> "Market":
        """Absolute vol bump: sigma -> sigma + size."""
        q = self.equity(name)
        eqs = dict(self.equities)
        eqs[name] = replace(q, vol=q.vol + size)
        return Market(self.curves, eqs, self.fx)

    def bump_fx(self, pair: str, rel: float) -> "Market":
        """Relative FX spot bump: S -> S * (1 + rel)."""
        s = self.fx_spot(pair)
        fx = dict(self.fx)
        fx[pair] = s * (1.0 + rel)
        return Market(self.curves, self.equities, fx)


def load_market(curves_csv: Path, spots_csv: Path) -> Market:
    """Load a Market from ``curves.csv`` (currency,tenor,zero_rate) and
    ``spots.csv`` (kind,name,spot,vol,div_yield,eq_bucket)."""
    by_ccy: Dict[str, List] = {}
    with open(curves_csv, newline="") as f:
        for row in csv.DictReader(f):
            by_ccy.setdefault(row["currency"], []).append(
                (float(row["tenor"]), float(row["zero_rate"]))
            )
    curves = {}
    for ccy, pts in by_ccy.items():
        pts.sort()
        curves[ccy] = Curve(tuple(p[0] for p in pts), tuple(p[1] for p in pts))
    if not curves:
        raise ValueError(f"load_market: no curves in {curves_csv}")

    equities: Dict[str, EquityQuote] = {}
    fx: Dict[str, float] = {}
    with open(spots_csv, newline="") as f:
        for row in csv.DictReader(f):
            kind = row["kind"]
            if kind == "equity":
                equities[row["name"]] = EquityQuote(
                    spot=float(row["spot"]),
                    vol=float(row["vol"]),
                    div_yield=float(row["div_yield"]),
                    bucket=row["eq_bucket"],
                )
            elif kind == "fx":
                fx[row["name"]] = float(row["spot"])
            else:
                raise ValueError(f"load_market: unknown kind '{kind}' in {spots_csv}")
    return Market(curves, equities, fx)

File: java/src/main/java/com/quant/frtb/Market.java
package com.quant.frtb;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * Immutable market snapshot: zero curves per currency, equity quotes per
 * name, FX spots per pair. All bump helpers return <em>new</em> Market
 * objects so bump-and-revalue sensitivities cannot leak state.
 *
 * @param curves   zero curve per currency
 * @param equities equity quote per name
 * @param fx       spot per FORDOM pair (e.g. "EURUSD")
 */
public record Market(Map<String, Curve> curves,
                     Map<String, EquityQuote> equities,
                     Map<String, Double> fx) {

    /**
     * Equity market data: spot, flat lognormal vol, dividend yield, SBM bucket.
     *
     * @param spot     spot price (&gt; 0)
     * @param vol      flat lognormal volatility (&ge; 0)
     * @param divYield continuous dividend yield
     * @param bucket   pinned SBM equity bucket id
     */
    public record EquityQuote(double spot, double vol, double divYield, String bucket) {
        public EquityQuote {
            if (!Double.isFinite(spot) || spot <= 0.0) {
                throw new IllegalArgumentException(
                        "EquityQuote: spot must be positive finite, got " + spot);
            }
            if (!Double.isFinite(vol) || vol < 0.0) {
                throw new IllegalArgumentException("EquityQuote: vol must be >= 0, got " + vol);
            }
            if (!Double.isFinite(divYield)) {
                throw new IllegalArgumentException("EquityQuote: divYield must be finite");
            }
        }
    }

    /** Curve for a currency; {@link IllegalArgumentException} when missing. */
    public Curve curve(String ccy) {
        Curve c = curves.get(ccy);
        if (c == null) {
            throw new IllegalArgumentException("Market: no curve for currency '" + ccy + "'");
        }
        return c;
    }

    /** Equity quote by name; {@link IllegalArgumentException} when missing. */
    public EquityQuote equity(String name) {
        EquityQuote q = equities.get(name);
        if (q == null) {
            throw new IllegalArgumentException("Market: no equity quote for '" + name + "'");
        }
        return q;
    }

    /** FX spot by pair; {@link IllegalArgumentException} when missing. */
    public double fxSpot(String pair) {
        Double s = fx.get(pair);
        if (s == null) {
            throw new IllegalArgumentException("Market: no FX spot for pair '" + pair + "'");
        }
        return s;
    }

    // ------------------- bump helpers (all return new Market objects) ------

    /** Market with one currency's curve replaced. */
    public Market withCurve(String ccy, Curve curve) {
        Map<String, Curve> c = new LinkedHashMap<>(curves);
        c.put(ccy, curve);
        return new Market(c, equities, fx);
    }

    /** Absolute bump of one curve node: {@code z(tenor) -> z(tenor) + size}. */
    public Market bumpCurveNode(String ccy, double tenor, double size) {
        return withCurve(ccy, curve(ccy).bumpedNode(tenor, size));
    }

    /** Absolute parallel bump of one currency's curve. */
    public Market bumpCurveParallel(String ccy, double size) {
        return withCurve(ccy, curve(ccy).bumpedParallel(size));
    }

    /** Relative equity spot bump: {@code S -> S * (1 + rel)}. */
    public Market bumpEquitySpot(String name, double rel) {
        EquityQuote q = equity(name);
        Map<String, EquityQuote> eqs = new LinkedHashMap<>(equities);
        eqs.put(name, new EquityQuote(q.spot() * (1.0 + rel), q.vol(), q.divYield(), q.bucket()));
        return new Market(curves, eqs, fx);
    }

    /** Absolute equity vol bump: {@code sigma -> sigma + size}. */
    public Market bumpEquityVol(String name, double size) {
        EquityQuote q = equity(name);
        Map<String, EquityQuote> eqs = new LinkedHashMap<>(equities);
        eqs.put(name, new EquityQuote(q.spot(), q.vol() + size, q.divYield(), q.bucket()));
        return new Market(curves, eqs, fx);
    }

    /** Relative FX spot bump: {@code S -> S * (1 + rel)}. */
    public Market bumpFx(String pair, double rel) {
        double s = fxSpot(pair);
        Map<String, Double> f = new LinkedHashMap<>(fx);
        f.put(pair, s * (1.0 + rel));
        return new Market(curves, equities, f);
    }

    // ------------------------------------------------------------- loading --

    /**
     * Load a Market from {@code curves.csv} (currency,tenor,zero_rate) and
     * {@code spots.csv} (kind,name,spot,vol,div_yield,eq_bucket).
     *
     * @throws IllegalArgumentException on schema errors or empty curve data
     */
    public static Market load(Path curvesCsv, Path spotsCsv) {
        Map<String, TreeMap<Double, Double>> byCcy = new LinkedHashMap<>();
        for (Map<String, String> row : readCsv(curvesCsv)) {
            byCcy.computeIfAbsent(row.get("currency"), k -> new TreeMap<>())
                    .put(Double.parseDouble(row.get("tenor")),
                            Double.parseDouble(row.get("zero_rate")));
        }
        Map<String, Curve> curves = new LinkedHashMap<>();
        for (Map.Entry<String, TreeMap<Double, Double>> e : byCcy.entrySet()) {
            double[] ts = e.getValue().keySet().stream().mapToDouble(Double::doubleValue).toArray();
            double[] rs = e.getValue().values().stream().mapToDouble(Double::doubleValue).toArray();
            curves.put(e.getKey(), new Curve(ts, rs));
        }
        if (curves.isEmpty()) {
            throw new IllegalArgumentException("Market.load: no curves in " + curvesCsv);
        }

        Map<String, EquityQuote> equities = new LinkedHashMap<>();
        Map<String, Double> fx = new LinkedHashMap<>();
        for (Map<String, String> row : readCsv(spotsCsv)) {
            String kind = row.get("kind");
            if ("equity".equals(kind)) {
                equities.put(row.get("name"), new EquityQuote(
                        Double.parseDouble(row.get("spot")),
                        Double.parseDouble(row.get("vol")),
                        Double.parseDouble(row.get("div_yield")),
                        row.get("eq_bucket")));
            } else if ("fx".equals(kind)) {
                fx.put(row.get("name"), Double.parseDouble(row.get("spot")));
            } else {
                throw new IllegalArgumentException(
                        "Market.load: unknown kind '" + kind + "' in " + spotsCsv);
            }
        }
        return new Market(curves, equities, fx);
    }

    /** Tiny header-keyed CSV reader (the bundled files have no quoting). */
    static List<Map<String, String>> readCsv(Path path) {
        List<String> lines;
        try {
            lines = Files.readAllLines(path);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        if (lines.isEmpty()) {
            throw new IllegalArgumentException("readCsv: empty file " + path);
        }
        String[] header = lines.get(0).split(",", -1);
        List<Map<String, String>> rows = new ArrayList<>();
        for (int i = 1; i < lines.size(); i++) {
            if (lines.get(i).isBlank()) {
                continue;
            }
            String[] cells = lines.get(i).split(",", -1);
            Map<String, String> row = new LinkedHashMap<>();
            for (int c = 0; c < header.length; c++) {
                row.put(header[c], c < cells.length ? cells[c] : "");
            }
            rows.add(row);
        }
        return rows;
    }
}
